include/dylib.hpp

```diff
@@ -215,77 +197,41 @@ class dylib {
     }
 
 protected:
-    native_handle_type m_handle{};
-#if defined(_WIN32) || defined(_WIN64)
+    native_handle_type m_handle{nullptr};
+
     static native_handle_type _open(const char *path) noexcept {
+#if defined(_WIN32) || defined(_WIN64)
         return LoadLibraryA(path);
+#else
+        return dlopen(path, RTLD_NOW | RTLD_LOCAL);
+#endif
     }
-    static FARPROC _get_symbol(native_handle_type lib, const char *name) noexcept {
-        return GetProcAddress(lib, name);
+
+    static DYLIB_WIN_MAC_OTHER(FARPROC, void*, void*)
+    _get_symbol(native_handle_type lib, const char *name) noexcept {
+        return DYLIB_WIN_OTHER(GetProcAddress, dlsym)(lib, name);
     }
+
     static void _close(native_handle_type lib) noexcept {
-        FreeLibrary(lib);
+        DYLIB_WIN_OTHER(FreeLibrary, dlclose)(lib);
     }
-    static char *_get_error() noexcept {
+
+    static std::string _get_error_description() noexcept {
+#if defined(_WIN32) || defined(_WIN64)
+        constexpr const size_t buf_size = 512;
         auto error_code = GetLastError();
         if (!error_code)
-            return nullptr;
-        char msg[512];
+            return "unknown error (GetLastError() failed)";
+        std::string description;
         auto lang = MAKELANGID(LANG_ENGLISH, SUBLANG_ENGLISH_US);
-        const DWORD len = FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM, nullptr, error_code, lang, msg, 512, nullptr);
-        if (len > 0)
-            return msg;
-        return nullptr;
-    }
+        const DWORD length =
+            FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM, nullptr, error_code, lang, description, buf_size, nullptr);
+        return (length > 0) ? description : nullptr;
```

Comment:
build an `std::string` with `nullptr` will result in a crash on linux

Reply:
Hmm, that's right... I need to replace that with an empty string.